src/ctap/hid/mod.rs

```diff
@@ -149,18 +151,16 @@ impl CtapHid {
     pub fn process_hid_packet<R, CheckUserPresence>(
         &mut self,
         packet: &HidPacket,
-        clock_value: ClockValue,
-        ctap_state: &mut CtapState<R, CheckUserPresence>,
+        clock_value: Instant<C>,
```

Comment:
I think we should provide helpers. The `clock` module should expose `Instant` as a simple type. The implementation using the actual clock as argument.

Reply:
I think since we don't expose the Instant outside the lib and we have the plan of using embedded_time::Timer, now we don't need to make this change. If with embedded_time:Timer we still need Instant, I'll change it in another PR.

Reply:
I'm not sure what you mean by "not exposed outside the lib" but I agree this can be changed in another PR (which would just add `pub type CtapInstant = Instant<CtapClock>` to the `clock` module). But at least in this PR it should be `Instant<CtapClock>`. The code shouldn't be generic on the clock since for a given compilation the clock is unique. That's the idea of the `clock` module to make the first PR simple. Long-term, when we really make CTAP a library, we would need to find the right abstractions (probably a single `Hardware` trait exposing other types). In other words here is how I see the overall "CTAP as a library" project:

Step 1: For each hardware component, move it to a module and only use what the module exposes. This should be small refactoring and aims at having clear interfaces. A typical module would look like:
```rust
// What is the API for this component?
pub trait Interface { ... } // could be provided by a library (like embedded_time)
// The wrapper is just to make sure we don't use anything outside the trait. (this could be step 1.b)
pub struct Wrapper<T>(T);
impl<T: Interface> Interface for Wrapper<T> { ... }
// What is the type for this component?
pub type Component<...> = Wrapper<Impl<...>>;  // could be directly Impl<...> for step1.a
// Some way to build the component
pub fn new(...) -> Component<...> { ... }
```

Step 2: Now that we have clear APIs, make the code generic over those APIs. To avoid too many generic parameters, we could use a single trait:
```rust
trait Hardware {
    type Clock: clock::Interface;  // the rest of the clock module is useless
    type Usb: usb::Interface;
    ...
}
```
Some functions/types would just be abstract over `Hardware`:
```rust
struct Foo<H: Hardware> {
    instant: Instant<H::Clock>,
}
```
Some would take some components as arguments:
```rust
fn foo<H: Hardware>(clock: &mut H::Clock, ...) { ... }
```

Reply:
I see. I tried to remove the generics, but it is a problem for the tests as they would then require the `CtapClock` which requires a `libtock_drivers::timer::Timer` to be created. How would you approach this problem?